### PR TITLE
AspNetCore Mvc logging

### PR DESCRIPTION
**Pull request by rossbuggins**
Initial start for logging AspNetCore Mvc values to Serilog through IDiagnosticContext

### CODE REVIEW

File: Serilog.AspNetCore.Mvc/Serilog.AspNetCore.Mvc.csproj [hunk 1]
```diff
@@ -0,0 +1,15 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Serilog" Version="2.8.0" />
+    <PackageReference Include="Serilog.Extensions.Hosting" Version="3.0.0" />
```
Comment (sungam3r) [inline Serilog.AspNetCore.Mvc/Serilog.AspNetCore.Mvc.csproj]: Serilog.Extensions.Hosting already depends from Serilog

File: Serilog.AspNetCore.Mvc/Serilog.AspNetCore.Mvc.csproj [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    <PackageReference Include="Microsoft.AspNetCore.Mvc.Core" Version="2.0.0" />
```
Comment (sungam3r) [inline Serilog.AspNetCore.Mvc/Serilog.AspNetCore.Mvc.csproj]: Microsoft.AspNetCore.Mvc.Core already depends from Microsoft.Extensions.DependencyInjection

File: Serilog.AspNetCore.Mvc/AspNetCore.Mvc/MvcLoggerFilter.cs [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        /// IDiagnosticContext by calling SetDiagFromValues.
+        /// </summary>
+        /// <param name="context"></param>
+        public override void OnActionExecuting(ActionExecutingContext context)
+        {
+            var values = new Dictionary<string, object>();
+            if (context == null)
+                return;
+
+            if(context.HttpContext!=null)
```
Comment (sungam3r) [inline Serilog.AspNetCore.Mvc/AspNetCore.Mvc/MvcLoggerFilter.cs]: whitespaces

File: Serilog.AspNetCore.Mvc/AspNetCore.Mvc/MvcLoggerFilter.cs [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+            {
+                LogHttpContextRequest(values, context.HttpContext);
+            }
+
+            if (context.ModelState.IsValid)
+            {
+
+                if(context!=null)
```
Comment (sungam3r) [inline Serilog.AspNetCore.Mvc/AspNetCore.Mvc/MvcLoggerFilter.cs]: whitespaces

File: Serilog.AspNetCore.Mvc/AspNetCore.Mvc/MvcLoggerFilter.cs [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+                {
+                    LogControllerExecutingAction(values, context);
+                }
+
+                var ctrlActionDesc = context.ActionDescriptor as ControllerActionDescriptor;
+                if(ctrlActionDesc!=null)
```
Comment (sungam3r) [inline Serilog.AspNetCore.Mvc/AspNetCore.Mvc/MvcLoggerFilter.cs]: whitespaces

File: Serilog.AspNetCore.Mvc/AspNetCore.Mvc/MvcLoggerFilter.cs [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        }
+
+        /// <summary>
+        /// Overrides the ActionFilterAttribute. Calls the built in log writers, compound and then finshes by writing to
+        /// IDiagnosticContext by calling SetDiagFromValues.
+        /// </summary>
+        /// <param name="context"></param>
+        public override void OnActionExecuting(ActionExecutingContext context)
+        {
+            var values = new Dictionary<string, object>();
```
Comment (sungam3r) [inline Serilog.AspNetCore.Mvc/AspNetCore.Mvc/MvcLoggerFilter.cs]: Please move this allocation below context check.

File: Serilog.AspNetCore.Mvc/AspNetCore.Mvc/MvcLoggerFilter.cs [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+        protected virtual void SetDiagFromValues(Dictionary<string, object> values)
+        {
+            foreach (var value in values)
+            {
+                diag.Set(value.Key, value.Value);
+            }
+        }
+
+        /// <summary>
+        /// Overrides the ActionFilterAttribute. Calls the built in log writers, compound and then finshes by writing to
```
Comment (sungam3r) [inline Serilog.AspNetCore.Mvc/AspNetCore.Mvc/MvcLoggerFilter.cs]: finshes

File: Serilog.AspNetCore.Mvc/AspNetCore.Mvc/MvcLoggerFilter.cs [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+        /// <param name="values"></param>
+        /// <param name="key"></param>
+        /// <param name="value"></param>
+        /// <returns></returns>
+        protected virtual Dictionary<string, object> Add(
+            Dictionary<string, object> values, 
+            string key, 
+            object value)
+        {
+            values.Add(key, value);
```
Comment (sungam3r) [inline Serilog.AspNetCore.Mvc/AspNetCore.Mvc/MvcLoggerFilter.cs]: error prone, consider `values[key] = value;`

File: Serilog.AspNetCore.Mvc/AspNetCore.Mvc/MvcLoggerFilter.cs [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+using System.Collections.Generic;
+
+namespace Serilog.AspNetCore.Mvc
+{
+    /// <summary>
+    /// Supports logging Asp.Net Core Mvc route information.
+    /// </summary>
+    public class MvcRequestLoggingFilter : ActionFilterAttribute
+    {
+        IDiagnosticContext diag;
```
Comment (sungam3r) [inline Serilog.AspNetCore.Mvc/AspNetCore.Mvc/MvcLoggerFilter.cs]: diag -> _diag

the project uses this naming style

**Comment by nblumhardt** [comment]
Thanks for the PR, Ross! This looks like a nice use of `IDiagnosticContext`.

My gut feel is that this is too broad and complex to include directly in _Serilog.AspNetCore_ in this form, but there's a tonne of useful stuff, here.

One option might be to spin up a separate repository to iterate on the design and publish some unofficial NUPKGs from, and once it's stabilised, look at including some of it here?

Minor note - dots in property names cause trouble for some Serilog sinks, it is generally better to create `StructureValue` objects with child properties in these cases.

**Comment by nblumhardt** [comment]
@rossbuggins I'd love to follow this along further, if you decide to spin up a repo/project/gist/etc. to develop it. Let me know how it turns out, and shout if there's anything I can do to help. Cheers!

**Comment by rossbuggins** [comment]
Hi ok, I’ll start this up next week and send through the details, cheers, Ross

________________________________

**Comment by rossbuggins** [comment]
I've created a repo at https://github.com/rossbuggins/serilog-aspnetcore-mvc

**Pull request closed without merging**